File: Script/parse_stats.py
#!/usr/bin/env python3
"""
parse_stats.py: Parse a .stats input (from a file or stdin) and display its contents in a formatted table.

Usage:
    # Read from a file:
    python parse_stats.py sup_exp.stats

    # Read from stdin:
    cat sup_exp.stats | python parse_stats.py

Dependencies:
    pip install tabulate
"""
import sys
import argparse
from tabulate import tabulate

def parse_stats(file_obj):
    """
    Read a .stats file-like object and extract metric, value, and optional comment.
    Lines starting with 'SN' are parsed; others are ignored.
    """
    rows = []
    for line in file_obj:
        line = line.strip()
        # Only process lines beginning with 'SN\t'
        if not line.startswith("SN\t"):
            continue
        parts = line.split("\t")
        # parts: ['SN', metric, value, comment?]
        metric = parts[1]
        value = parts[2]
        comment = ""
        if len(parts) > 3:
            # Remove leading '#' if present
            comment = parts[3].lstrip('#').strip()
        rows.append((metric, value, comment))
    return rows

def main():
    parser = argparse.ArgumentParser(
        description="Parse a .stats input (file or stdin) into a formatted table"
    )
    parser.add_argument(
        "stats_file",
        nargs="?",
        type=argparse.FileType('r'),
        default=sys.stdin,
        help="Path to the .stats file; if omitted, reads from stdin"
    )
    args = parser.parse_args()

    # Parse rows from the provided file or stdin
    rows = parse_stats(args.stats_file)
    if not rows:
        print("No valid 'SN' entries found in the input.")
        return

    # Display the table with custom alignment
    print('\nAlignment Statistics:\n')
    print(tabulate(
        rows,
        headers=["Metric", "Value", "Comment"],
        tablefmt="fancy_grid",
        colalign=("left", "right", "left")
    ))

if __name__ == "__main__":
    main()

File: src/c++/ParseStats.cpp
// ParseStats.cpp
// Produces executable: ParseStats
// Usage: ParseStats [stats_file]   (stdin if omitted)
//
// Pipeline role: pretty-print the 'SN' (summary-numbers) section of a
// samtools stats output as a three-column table (Metric, Value, Comment).
// Used to embed alignment summaries in the thesis reports without pulling
// in awk/cut boilerplate in every analysis script.
#include <fstream>
#include <iomanip>
#include <iostream>
#include <string>
#include <vector>

using namespace std;

// Single parsed 'SN' row; comment is optional and has any leading '# ' trimmed.
struct Row {
    string metric;
    string value;
    string comment;
};

// Print CLI contract to stderr when args are invalid.
static void usage(const char* prog) {
    cerr << "Usage: " << prog << " [stats_file]\n\n"
         <<
R"(Usage:
Description:
  Parse samtools-style .stats input and print the SN lines in a formatted table.
  If no file is provided, input is read from stdin.
)";
}

// Read the stream and return only the 'SN' summary rows, split on tabs.
// Skips every other section emitted by samtools stats (FFQ, GCF, IS, ...).
static vector<Row> parse_rows(istream& in) {
    vector<Row> rows;
    string line;
    while (getline(in, line)) {
        // Keep only lines that start with the SN section prefix.
        if (line.rfind("SN\t", 0) != 0) continue;

        vector<string> fields;
        string token;
        size_t start = 0;
        while (true) {
            size_t pos = line.find('\t', start);
            if (pos == string::npos) {
                fields.push_back(line.substr(start));
                break;
            }
            fields.push_back(line.substr(start, pos - start));
            start = pos + 1;
        }

        if (fields.size() < 3) continue;
        Row row;
        row.metric = fields[1];
        row.value = fields[2];
        // An optional 4th field holds samtools' free-text comment, which
        // usually begins with '#'; strip the '#' and any leading spaces
        // so the comment column in the output is plain English.
        if (fields.size() > 3) {
            row.comment = fields[3];
            if (!row.comment.empty() && row.comment[0] == '#') {
                row.comment.erase(0, 1);
                while (!row.comment.empty() && row.comment[0] == ' ') row.comment.erase(0, 1);
            }
        }
        rows.push_back(std::move(row));
    }
    return rows;
}

int main(int argc, char** argv) {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    if (argc > 2) {
        usage(argv[0]);
        return 1;
    }
    if (argc == 2) {
        string arg = argv[1];
        if (arg == "--help" || arg == "-h") {
            usage(argv[0]);
            return 0;
        }
    }

    istream* in = &cin;
    ifstream file;
    if (argc == 2) {
        file.open(argv[1]);
        if (!file) {
            cerr << "Error: cannot open '" << argv[1] << "'\n";
            return 2;
        }
        in = &file;
    }

    vector<Row> rows = parse_rows(*in);
    if (rows.empty()) {
        cout << "No valid 'SN' entries found in the input.\n";
        return 0;
    }

    // Compute per-column widths so the output lines up regardless of the
    // longest metric name or value in the current stats file.
    size_t metric_w = string("Metric").size();
    size_t value_w = string("Value").size();
    size_t comment_w = string("Comment").size();
    for (const auto& row : rows) {
        metric_w = max(metric_w, row.metric.size());
        value_w = max(value_w, row.value.size());
        comment_w = max(comment_w, row.comment.size());
    }

    cout << "\nAlignment Statistics:\n\n";
    cout << left << setw(static_cast<int>(metric_w)) << "Metric" << "  "
         << right << setw(static_cast<int>(value_w)) << "Value" << "  "
         << left << setw(static_cast<int>(comment_w)) << "Comment" << "\n";
    cout << string(metric_w, '-') << "  "
         << string(value_w, '-') << "  "
         << string(comment_w, '-') << "\n";

    for (const auto& row : rows) {
        cout << left << setw(static_cast<int>(metric_w)) << row.metric << "  "
             << right << setw(static_cast<int>(value_w)) << row.value << "  "
             << left << setw(static_cast<int>(comment_w)) << row.comment << "\n";
    }

    return 0;
}
